starter-code/app.js

```diff
@@ -1,5 +1,57 @@
-const express = require('express');
+const express = require("express");
 const app = express();
-const Chuck  = require('chucknorris-io');
+const Chuck = require("chucknorris-io");
 const client = new Chuck();
 
+// app.use(expressLayouts);
+// app.set("layout", "layouts/main-layout");
+app.set("views", __dirname + "/views");
+app.set("view engine", "ejs");
+
+app.get("/", (request, response, next) => {});
+
+app.get("/random", (req, res, next) => {
+  client
+    .getRandomJoke()
+    .then(result => {
+      //console.log(result);
+      res.send(`<p>${result.value}</p>`);
+    })
+    .catch(err => {
+      throw err;
+    });
+});
+
+app.get("/categories", (req, res, next) => {
+  client
+    .getJokeCategories()
+    .then(result => {
+      var jokeCategories = [];
+      res.render("joke-by-category", {
+        jokeCategories: result
+      });
+      console.log(jokeCategories);
+    })
+    .catch(err => {
+      throw err;
+    });
+});
+
+for (i = 0; i < jokeCategories.length; i++) {
```

Comment:
This is causing your server to crash, since `jokeCategories` is a local variable in the scope of one of the callbacks in the former route handler. In this global scope, this variable doesn't exist.

Reply:
You are trying to create a dynamic route handler. Instead of putting it in a loop, Express provides a way to do that easily. Check out https://expressjs.com/en/guide/routing.html in the *Route parameters* section.

Reply:
And let me know if you have questions. In any case, we'll go deeper into this later.